### PR TITLE
Add support for browserslist and legacyBrowsers experimental option

### PR DESCRIPTION
**Pull request by timneutkens**
Implements the first part of #33227

- Applies browserslist to JS transforms when `experimental.browsersListForSwc` is enabled. 
- You don't have to use browserslist, there's also `legacyBrowsers: false` which will be the new default in Next.js 13. See #33227 for which browsers and why. `legacyBrowsers` requires `browsersListForSwc: true` to function until it is the default. 

```js
module.exports = {
  experimental: {
    legacyBrowsers: false,
    browsersListForSwc: true,
  }
}
```

I only implemented the JS part of the RFC, the CSS part should be handled in a follow-up PR.

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [x] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [x] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 8a14e1740b841218519a53b7fa8b82250a90a5b5

`yarn testheadless test/e2e/browserslist-extends/index.test.ts` 
- browserslist-extends > should work

<details>
<summary>Expand output</summary>

  ● browserslist-extends › should work

    expect(received).toContain(expected) // indexOf

    Expected substring: "hello world"
    Received string:    "Internal Server Error"

      34 |   it('should work', async () => {
      35 |     const html = await renderViaHTTP(next.url, '/')
    > 36 |     expect(html).toContain('hello world')
         |                  ^
      37 |   })
      38 | })
      39 |

      at Object.<anonymous> (e2e/browserslist-extends/index.test.ts:36:18)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| buildDuration | 19.2s | 19.1s | -44ms |
| buildDurationCached | 7.7s | 7.5s | -126ms |
| nodeModulesSize | 463 MB | 463 MB | ⚠️ +1.82 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.518 | 4.513 | 0 |
| / avg req/sec | 553.38 | 553.92 | +0.54 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.629 | 1.615 | -0.01 |
| /error-in-render avg req/sec | 1534.81 | 1547.78 | +12.97 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.3 kB | 72.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.71 kB | 5.71 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.64 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.3 kB | 16.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 528 B | 528 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| buildDuration | 21.5s | 21.2s | -241ms |
| buildDurationCached | 7.8s | 7.7s | -43ms |
| nodeModulesSize | 463 MB | 463 MB | ⚠️ +1.82 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.534 | 4.503 | -0.03 |
| / avg req/sec | 551.43 | 555.21 | +3.78 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.64 | 1.715 | ⚠️ +0.08 |
| /error-in-render avg req/sec | 1524.61 | 1458 | ⚠️ -66.61 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.1 kB | 29.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.9 kB | 72.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.76 kB | 2.76 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.5 kB | 16.5 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/legacybrowsers-rfc | Change |
| - | - | - | - |
| index.html gzip | 529 B | 529 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 8a14e1740b841218519a53b7fa8b82250a90a5b5

**Review by SukkaW — COMMENTED** [review]
Isn't it confusing that while swc / babel does transform the modern syntax without including the corresponding polyfill?

E.g. If I set my `browserslists` to `Chrome > 64` and use `Promise.allSettled` in my code. I would be expecting the corresponding polyfill should be included because `Promise.allSettled` is only supported since Chrome 76.

**Comment by timneutkens** [comment]
> Isn't it confusing that while swc / babel does transform the modern syntax without including the polyfill?

What's the problem you're referring to?

**Comment by timneutkens** [comment]
> Isn't it confusing that while swc / babel does transform the modern syntax without including the corresponding polyfill?
> 
> E.g. If I set my `browserslists` to `Chrome > 64` and use `Promise.allSettled` in my code. I would be expecting the corresponding polyfill should be included because `Promise.allSettled` is only supported since Chrome 76.

Next.js has 2 sets of polyfills that are loaded which are based on module/nomodule support. In case of `Promise.allSettled` I think it should be added to `next-polyfill-module` as it's a common feature.

**Comment by SukkaW** [comment]
> > Isn't it confusing that while swc / babel does transform the modern syntax without including the corresponding polyfill?
> > E.g. If I set my `browserslists` to `Chrome > 64` and use `Promise.allSettled` in my code. I would be expecting the corresponding polyfill should be included because `Promise.allSettled` is only supported since Chrome 76.
> 
> Next.js has 2 sets of polyfills that are loaded which are based on module/nomodule support. In case of `Promise.allSettled` I think it should be added to `next-polyfill-module` as it's a common feature.

IMHO it is a really bad idea. With the ECMAScript rolling out the spec every year, the size of the `next-polyfill-module` will continuously increase. And how can you tell whether an ECMAScript feature is commonly used by a Next.js user or not? E.g. I use `Object.fromEntries` in my project but Next.js refuses to add it to the polyfill because it will impact the size of the `next-polyfill-module` (And it really confuses people as they believe Next.js is full ES2019 compatible. And that's why I create #36426, an `Object.fromEntries` polyfill within 122 bytes).

And what happened to those who set `browserlists` to `chrome > 100`? Maybe they don't want any polyfill and transformation at all, maybe they are just prototyping an idea with SSG/SSR, but they still need to include `next-plyfill-module` in their bundle. Not a very good idea.

**Comment by timneutkens** [comment]
@SukkaW the polyfill bundle is not the subject of the RFC #33227, the RFC only covers changing the defaults for transforming JS/CSS and not the polyfilling behavior, hence why it's not covered in this PR. Feel free to open a new RFC to talk about changing the defaults for polyfilling with the points mentioned. E.g. a RFC to change polyfilling based on the browserslist configuration could be considered a way to go after these changes land. There's a bunch of historical context / reasons why we went with the module/nomodule polyfill bundles approach over having Babel inject core-js polyfills like that it actually **reduced** bundle size for all applications when it was introduced so dismissing it as "a really bad idea" is not really constructive. Here's the data on size impact for **all apps** by default: https://github.com/vercel/next.js/pull/10212.

**Comment by SukkaW** [comment]
> For example that it actually **reduced** bundle size for all applications when it was introduced so dismissing it as "a really bad idea" is not really constructive.

The reason behind the final bundle size reduction is because it is not how `core-js` works. `core-js` not only provides a polyfill for the unsupported features but also fixes bugs for modern browsers' wrong implementation.

For instance, core-js provided a workaround for a critical bug about `Array.prototype.reduce` introduced in Chrome 80-82: https://github.com/zloirock/core-js/blob/ce52fdc735c5c809c9e85b2072f92d41b5a3885a/packages/core-js/modules/es.array.reduce.js

And another example, core-js provided a workaround for a critical bug about `Array.prototype.reverse` introduced in Safari 12:
https://github.com/zloirock/core-js/blob/fcbc2c878535b277f840997ab8ee819efd677888/packages/core-js/modules/es.array.reverse.js

File: test/e2e/browserslist/browserslist.test.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    let finished = false
+    await Promise.all(
+      $('script')
+        .toArray()
+        .map(async (el) => {
+          const src = $(el).attr('src')
+          if (!src) return
+          if (src.includes('/index')) {
+            const source = next.url + src
+            const code = await fetch(source).then((res) => res.text())
```
Comment (ijjk) [inline test/e2e/browserslist/browserslist.test.ts:36]:
```suggestion
            const code = await fetchViaHTTP(next.url, source).then((res) => res.text())
```

File: test/e2e/browserslist/legacybrowsers-false.test.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+          legacyBrowsers: false,
+          browsersListForSwc: true,
+        },
+      },
+      dependencies: {},
+    })
+  })
+  afterAll(() => next.destroy())
+
+  it('should apply legacyBrowsers: true by default', async () => {
```
Comment (ijjk) [inline test/e2e/browserslist/legacybrowsers-false.test.ts:27]:
```suggestion
  it('should apply with legacyBrowsers: false correctly', async () => {
```

**Review by ijjk — APPROVED** [review]

**Review by styfle — COMMENTED** [review]
Should we add docs to explain how to configure this?

**Pull request merged**